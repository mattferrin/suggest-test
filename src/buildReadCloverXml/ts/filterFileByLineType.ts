import { MeowOutput } from "../../bin/ts/cli";
import { FileArray } from "./summarizedResult";
import { Untrusted } from "./Untrusted";

export function filterFileByLineType(
  input: MeowOutput,
  projectPackage: Untrusted<{
    readonly file: FileArray;
  }>
): Untrusted<FileArray> {
  if (!input.flags.includeStatements) {
    return projectPackage?.file?.filter((projectPackageFile) => {
      return projectPackageFile?.line?.some((line) => {
        return line?.$?.type === "cond";
      });
    });
  } else {
    return projectPackage?.file;
  }
}
