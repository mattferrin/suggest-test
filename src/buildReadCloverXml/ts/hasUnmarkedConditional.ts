import { MeowOutput } from "../../bin/ts/cli";
import { FileArray } from "./summarizedResult";
import { Untrusted } from "./Untrusted";

export function hasUnmarkedConditional(
  input: MeowOutput,
  file: Untrusted<FileArray[number]>
): boolean {
  if (input.flags.includeConditionals) {
    const hasConditional =
      file?.metrics?.some((metric) => {
        return Number.parseInt(metric?.$?.conditionals ?? "0") > 0;
      }) ?? false;

    return (
      hasConditional &&
      (file?.line?.every((line) => {
        return line?.$?.type !== "cond";
      }) ??
        false)
    );
  } else {
    return false;
  }
}
