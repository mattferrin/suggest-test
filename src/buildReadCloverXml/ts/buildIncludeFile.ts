import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function buildIncludeFile(input: MeowOutput) {
  return function includeFile(file: SummaryItem | undefined) {
    const isTested =
      (file?.statements ?? 0) > (file?.uncoveredStatements ?? 0) ||
      (file?.conditionals ?? 0) > (file?.uncoveredConditionals ?? 0);
    const isIncludeTested = input.flags.includeTested && isTested;

    const isNotTested =
      (file?.statements ?? 0) === (file?.uncoveredStatements ?? 0) &&
      (file?.conditionals ?? 0) === (file?.uncoveredConditionals ?? 0);
    const isIncludeNotTested = input.flags.includeNotTested && isNotTested;

    return isIncludeTested || isIncludeNotTested;
  };
}
