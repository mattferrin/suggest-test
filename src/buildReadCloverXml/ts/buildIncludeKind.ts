import { MeowOutput } from "../../bin/ts/cli";
import { SummaryItem } from "./undefinedOrSummary";

export function buildIncludeKind(input: MeowOutput) {
  return function includeKind(file: SummaryItem | undefined) {
    const isStatementKeeper =
      (file?.statements ?? 0) > 1 && (file?.uncoveredStatements ?? 0) > 0;
    const isConditionalKeeper =
      (file?.conditionals ?? 0) > 1 && (file?.uncoveredConditionals ?? 0) > 0;

    if (!input.flags.includeConditionals) {
      return isStatementKeeper;
    } else if (!input.flags.includeStatements) {
      return isConditionalKeeper;
    } else {
      return isStatementKeeper || isConditionalKeeper;
    }
  };
}
