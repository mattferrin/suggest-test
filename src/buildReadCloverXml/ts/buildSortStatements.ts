import { MeowOutput } from "../../bin/ts/cli";
import { ReadCloverXmlOutput } from "./buildReadCloverXml";
import { summarizeIntoString } from "./summarizeIntoString";

export function buildSortStatements(input: MeowOutput) {
  return function sortStatements(
    output: ReadCloverXmlOutput
  ): ReadCloverXmlOutput {
    switch (output.tag) {
      case "errors":
        return output;
      case "summary": {
        const summaryArray = output.summaryArray
          ?.filter((file) => (file?.statements ?? 0) > 1)
          ?.sort((fileA, fileB) => {
            if (
              fileB?.statements === null ||
              fileB?.statements === undefined ||
              fileA?.statements === null ||
              fileA?.statements === undefined
            ) {
              return -1;
            } else {
              return fileA.weightedStatements > fileB.weightedStatements
                ? -1
                : 1;
            }
          });

        return {
          tag: "summary",
          summaryArray,
          summaryString: summarizeIntoString(summaryArray),
        };
      }
    }
  };
}
