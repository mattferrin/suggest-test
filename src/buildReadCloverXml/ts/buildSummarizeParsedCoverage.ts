import { MeowOutput } from "../../bin/ts/cli";
import { filterFileByLineType } from "./filterFileByLineType";
import { FileArray, summarizedResult } from "./summarizedResult";
import { SummaryItem } from "./undefinedOrSummary";
import { Untrusted } from "./Untrusted";

export type Parsed = Untrusted<{
  readonly coverage: {
    readonly project: ReadonlyArray<{
      readonly package: ReadonlyArray<{
        readonly file: FileArray;
      }>;
    }>;
  };
}>;

type SummarizeParsedCoverageOutput =
  | ReadonlyArray<SummaryItem | undefined>
  | undefined;

export function buildSummarizeParsedCoverage(input: MeowOutput) {
  return function summarizeParsedCoverage(
    parsed: Parsed
  ): SummarizeParsedCoverageOutput {
    return parsed?.coverage?.project?.flatMap((project) => {
      return project?.package?.flatMap((projectPackage) => {
        const filteredProjectPackageFile = filterFileByLineType(
          input,
          projectPackage
        );

        return summarizedResult(input, filteredProjectPackageFile);
      });
    });
  };
}
