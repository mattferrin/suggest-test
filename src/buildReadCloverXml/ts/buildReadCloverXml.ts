import fs from "fs/promises";
import { parseStringPromise } from "xml2js";
import { MeowOutput } from "../../bin/ts/cli";
import { buildSortStatements } from "./buildSortStatements";
import { buildSummarizeParsedCoverage } from "./buildSummarizeParsedCoverage";
import { passThroughLogger } from "./passThroughLogger";
import { SummaryItem, undefinedOrSummary } from "./undefinedOrSummary";

export type ReadCloverXmlOutput =
  | { readonly tag: "errors"; readonly errors: readonly Error[] }
  | {
      readonly tag: "summary";
      readonly summaryArray: ReadonlyArray<SummaryItem | undefined>;
      readonly summaryString: string | { readonly tag: "not-ready" };
    };

export function buildReadCloverXml(input: MeowOutput) {
  // eslint-disable-next-line functional/functional-parameters
  return async function readCloverXml(): Promise<ReadCloverXmlOutput> {
    return await fs
      .readFile(input.input[0] ?? "coverage/clover.xml")
      .then(parseStringPromise)
      .then(buildSummarizeParsedCoverage(input))
      .then(undefinedOrSummary)
      .then(buildSortStatements(input))
      .catch((reason): ReadCloverXmlOutput => {
        return { tag: "errors", errors: [new Error(reason)] };
      })
      .then(passThroughLogger);
  };
}
