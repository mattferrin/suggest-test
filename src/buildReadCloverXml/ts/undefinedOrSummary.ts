import { ReadCloverXmlOutput } from "./buildReadCloverXml";

export interface SummaryItem {
  readonly conditionals: number | null | undefined;
  readonly path: string | null | undefined;
  readonly statements: number | null | undefined;
  readonly uncoveredConditionals: number;
  readonly uncoveredStatements: number;
  readonly score: number;
  readonly hasUnmarkedConditional: boolean;
}

export function undefinedOrSummary(
  summary: ReadonlyArray<SummaryItem | undefined> | undefined
): ReadCloverXmlOutput {
  if (summary === undefined || summary === null) {
    return {
      tag: "errors",
      errors: [
        new Error(
          "The parsed clover.xml has an undefined or null property somewhere"
        ),
      ],
    };
  } else {
    return {
      tag: "summary",
      summaryArray: summary,
      summaryString: { tag: "not-ready" },
    };
  }
}
