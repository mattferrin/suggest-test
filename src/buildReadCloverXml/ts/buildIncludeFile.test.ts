import { MeowOutput } from "../../bin/ts/cli";
import { buildIncludeFile } from "./buildIncludeFile";
import { SummaryItem } from "./undefinedOrSummary";

function snap(
  uuid: string,
  comment: string | null,
  input: MeowOutput,
  file: SummaryItem | undefined
): void {
  it(uuid, () => {
    /** setup mocks */

    /** unit under test */
    const includeFile = buildIncludeFile(input);
    const result = includeFile(file);

    /** assertions */
    expect({
      [`${comment ?? "TODO: comment"}`]: { result },
    }).toMatchSnapshot();
  });
}

snap(
  "5c9c07a3-5e29-45fe-af63-8780df887809",
  "undefined file doesn't throw an error WHEN includeNotTested false",
  { flags: { includeTested: false, includeNotTested: false } } as MeowOutput,
  undefined
);

snap(
  "a036467a-aa43-438a-a54a-220cbb3539fb",
  "undefined file doesn't throw an error WHEN includeNotTested true",
  { flags: { includeTested: false, includeNotTested: true } } as MeowOutput,
  undefined
);

snap(
  "5bf05b28-2df7-43a9-944b-72c65acccbb9",
  "returns `true` for file because at least 1 statement is tested",
  { flags: { includeTested: true, includeNotTested: false } } as MeowOutput,
  { statements: 2, uncoveredStatements: 1 } as SummaryItem
);

snap(
  "8d1e30e7-4f17-4fee-a36c-4da484d0e25e",
  "returns `true` for file because at least 1 conditional is tested",
  { flags: { includeTested: true, includeNotTested: false } } as MeowOutput,
  { conditionals: 2, uncoveredConditionals: 1 } as SummaryItem
);

snap(
  "13a1c1c6-dd44-4fc4-8102-da679ba19713",
  "returns `true` for file because every line and conditional is not tested",
  { flags: { includeTested: false, includeNotTested: true } } as MeowOutput,
  {
    conditionals: 9,
    uncoveredConditionals: 9,
    statements: 99,
    uncoveredStatements: 99,
  } as SummaryItem
);
