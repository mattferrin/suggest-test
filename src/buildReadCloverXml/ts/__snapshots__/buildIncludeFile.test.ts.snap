// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`5bf05b28-2df7-43a9-944b-72c65acccbb9 1`] = `
{
  "returns \`true\` for file because at least 1 statement is tested": {
    "result": true,
  },
}
`;

exports[`5c9c07a3-5e29-45fe-af63-8780df887809 1`] = `
{
  "undefined file doesn't throw an error WHEN includeNotTested false": {
    "result": false,
  },
}
`;

exports[`8d1e30e7-4f17-4fee-a36c-4da484d0e25e 1`] = `
{
  "returns \`true\` for file because at least 1 conditional is tested": {
    "result": true,
  },
}
`;

exports[`13a1c1c6-dd44-4fc4-8102-da679ba19713 1`] = `
{
  "returns \`true\` for file because every line and conditional is not tested": {
    "result": true,
  },
}
`;

exports[`a036467a-aa43-438a-a54a-220cbb3539fb 1`] = `
{
  "undefined file doesn't throw an error WHEN includeNotTested true": {
    "result": true,
  },
}
`;
